### ISSUE TITLE
discussing

### ISSUE THREAD

**Issue by leyuan-sun** [utn-air/flownav#5]
Hi, I would like to ask about the mse loss `flow_loss = action_reduce(F.mse_loss(vt, ut, reduction="none"), action_mask)`, when robot execute the action, it can only move to the first predicted way point in this vt, so why we calcuate the mse loss on all the waypoints? Or why the all 8 points are averagre calculated the mse loss with GT position? Maybe I am wrong, but want to know your opinions， thanks!

**Comment by debOliveira** [comment]
We don't always move to the first waypoint. In FlowNav we follow NoMaD's velocity control, where the robot is direct to the third waypoint (index 2). You can chose which waypoint to go using the following argument:
https://github.com/utn-air/flownav/blob/0ae0632005c8563a84021d39df5381bccb2ce7ca/deployment/src/navigation/navigate.py#L270-L276

Diffusion policies uses a long sequence horizon to improve temporal stability and consistency. You can check more details on it on [their paper](https://diffusion-policy.cs.columbia.edu/diffusion_policy_2023.pdf).

**Comment by leyuan-sun** [comment]
Thanks for your information, in your/nomad deployment, actually robot just moves to the third waypoint and then the model inference the next 8 future actions/points, which is different from your attachted paper that "which Ta steps of actions are executed on the robot without re-planning". They used the next T_a steps actions without re-planning. Why you dont use the next i.e. 4 points instead of only using one point once. Besides,why the all 8 points are averagre calculated the mse loss with GT position between vt and ut?  I am still confused about this part. Thanks for sharing your opinions.

**Comment by leyuan-sun** [comment]
Sry, one more question, there are many sampled trajectories generateed by nomad or flownav, duing real world deployment, how does the robot select from them, random?

**Comment by abhijeetknayak** [comment]
Regarding your first comment, yes you could also use each waypoint from the first Ta waypoints to define the linear and angular velocities. We haven't tried this ourselves yet, but if you do, you could let us know how the robot performs in the real world.
However, a simpler way to execute these trajectories would be to choose a waypoint and just make a step in that direction for a small t, after which a new action is generated and executed (you could call this re-planning). This is what the previous state-of-the-art methods (NoMaD, ViNT and GNM) do - see [here](https://github.com/robodhruv/visualnav-transformer/tree/main)

The 8 points are used to make sure a full, smooth trajectory is learnt, rather than single points. Denoising a full trajectory is more stable and robust, which is why most state-of-the-art diffusion policies for navigation or manipulation use a bigger prediction horizon and a smaller execution horizon.

Regarding your next comment, an action is chosen at random. Therefore, we always choose the action at the 0th index. You could also select a random index and use that action. Better selection strategies can be defined but that can be part of future work.

**Comment by debOliveira** [comment]
> Thanks for your information, in your/nomad deployment, actually robot just moves to the third waypoint and then the model inference the next 8 future actions/points, which is different from your attachted paper that "which Ta steps of actions are executed on the robot without re-planning". **They used the next T_a steps actions without re-planning. Why you dont use the next i.e. 4 points instead of only using one point once.** 

To complement @abhijeetknayak , we use next 4 waypoints as is in our [newest research](https://utn-air.github.io/metricnet/). First, we anchor the hand-tuned scale to metric units. For manipulation, the policy must be fine-tuned to the specific embodiment. For navigation, calibration is done using the maximum linear and angular velocities. With NoMaD and related policies, the goal is to react quickly enough to correct the non-metric trajectory, so selecting a single waypoint matters less.
